corehq/sql_db/connections.py

```diff
@@ -100,20 +108,31 @@ def get_engine(self, engine_id=DEFAULT_ENGINE_ID):
         """
         return self._get_or_create_helper(engine_id).engine
 
-    def get_load_balanced_read_db(self, engine_id, default=None):
+    def get_load_balanced_read_engine_id(self, engine_id, default=None):
         read_dbs = self.read_database_mapping.get(engine_id, [])
-        if read_dbs:
-            return random.choice(read_dbs)
-        elif default is not None:
+        # convert to a db to weight dictionary
+        weights_by_db = {db: weight for db, weight in read_dbs}
+
+        # filter out stale standby dbs
+        fresh_dbs = filter_out_stale_standbys(
+            weights_by_db.keys(), ACCEPTABLE_STANDBY_DELAY_SECONDS
+        )
+        dbs = []
+        weights = []
+        for db, weight in six.iteritems(weights_by_db):
+            if db in fresh_dbs:
+                dbs.append(db)
+                weights.append(weight)
+
+        if dbs:
+            # normalize weights of remaining dbs
+            total_weight = sum(weights)
+            normalized_weights = [weight / total_weight for weight in weights]
+            return random.choice(dbs, p=normalized_weights)
```

Comment:
Rather than generating the dbs and weights on every request we could cache them. So rather than caching the `get_replication_delay_for_standby` we could cache this instead.

Reply:
Sure, is that recommended because the random number picking will take consider amount of time?